bip-0360.mediawiki

```diff
@@ -0,0 +1,724 @@
+ABCDE = tagged_hash("TapBranch", AB + CDE)
+Root = tagged_hash("QuantumRoot", ABCDE)
+</source>
+
+A P2QRH input witness provides the following:
+
+<source>
+initial stack element 0,
+...,
+initial stack element N,
+tapleaf script,
+control block = [control byte, 32 * m byte Merkle path] # m is the depth of the Merkle tree
+</source>
+
+The initial stack elements provide the same functionality as they do in P2TR. That is, they place elements on the stack to
+be evaluated by the script, a.k.a. the redeem script.
+
+The control block is a 1 + 32 * m byte array, where the first byte is the control byte and the next 32*m bytes are the
+Merkle path to the script. The control byte is the same as the control byte in a P2TR control block, 
+including the 7 bits which are used to specify the tapleaf version. The parity bit of the control byte is always 1
+since P2QRH does not have a key path spend. We omit the public key from the control block as it is not needed in P2QRH.
+We maintain support for the optional annex in the witness (see specification for more details).
+
+=== Rationale ===
+
+Our design to augment Bitcoin with quantum resistance is guided by the following principles:
+
+'''Minimize changes.''' We should reuse existing Bitcoin code and preserve 
+existing software behavior, workflows, user expectations and compatibility whenever possible.
+
+'''Gradual upgrade path.''' We should provide an upgrade path for wallets and exchanges which can be 
+carried out gradually and iteratively rather than all at once. This is critical as the earlier the ecosystem
+begins upgrading to quantum resistance, the lower the number of coins at risk when quantum attacks become practical.
+
+'''Use standardized post-quantum signature algorithms.'''  Standardized algorithms have undergone the most scrutiny and 
+are likely to be most well supported and well studied going forward. The entire Bitcoin ecosystem will benefit
+from using the most popular post-quantum signature algorithms, including leveraging hardware acceleration 
+instructions, commodity trusted hardware, software libraries and cryptography research.
+
+'''Provide security against unexpected cryptanalytic breakthroughs.''' Consider the risk
+if Bitcoin only supported one PQ signature algorithm, and then following the widespread rollout of CRQCs, a critical
+weakness is unexpectedly discovered in this signature algorithm. There would be no safe algorithm available. We believe that
+prudence dictates we take such risks seriously and ensure that Bitcoin always has at least two secure signature algorithms built
+on orthogonal cryptographic assumptions. In the event one algorithm is broken, an alternative will be available. An added benefit
+is that parties seeking to securely store Bitcoin over decades can lock their coins under multiple algorithms,
+ensuring their coins will not be stolen even in the face of a catastrophic break in one of those signature algorithms.
```

Comment:
_A contrario_, this is more cryptographic consensus code that have to be integrated, where any implementation bug might be fatal….

Personally, I think it’s worth the risk to have multiple signature schemes in case of advances in post-quantum cryptanalysis (it’s all quite unchartered paths here…), but I believe it might be more debatable among the community and the industry…